graspy/embed/oosase.py

```diff
@@ -0,0 +1,416 @@
+            else:
+                self.in_sample_idx = np.random.choice(
+                    self._nodes, int(N * self.in_sample_proportion)
+                )
+        else:
+            self.in_sample_proportion = len(self.in_sample_idx) / N
+
+        self._connected_subgraph(graph)
+
+        in_sample_G = graph.subgraph(self.in_sample_idx).copy()
+
+        in_sample_A = nx.to_numpy_array(in_sample_G)
+
+        self.id_to_index = dict(zip(self.in_sample_idx, range(len(self.in_sample_idx))))
+        self.index_to_id = dict(zip(range(len(self.in_sample_idx)), self.in_sample_idx))
+
+        self._reduce_dim(in_sample_A)
+
+        if self._directed:
+            self.in_sample_embedding = np.concatenate(
+                (self.latent_left_, self.latent_right_), axis=1
+            )
+        else:
+            self.in_sample_embedding = self.latent_left_
+
+        return self
+
+    def _connected_subgraph(self, graph):
+        """
+        Find a connected subgraph.
+
+        Parameters
+        ----------
+        graph : array_like or networkx.Graph
+            Input graph to embed.
+
+        Returns
+        -------
+        None
+
+        """
+        N = len(graph)
+        in_sample_G = graph.subgraph(self.in_sample_idx).copy()
+
+        if is_fully_connected(in_sample_G):
+            return
+
+        if self.in_sample_proportion < 1:
+            c = 0
+            less_than_half = True
+            while (
+                not (is_fully_connected(in_sample_G) and c < self.connected_attempts)
+                and less_than_half
+            ):
+                self.in_sample_idx = np.random.choice(
+                    self._nodes, int(N * self.in_sample_proportion)
+                )
+                temp = graph.subgraph(self.in_sample_idx).copy()
+                temp_lcc = max(nx.connected_components(temp), key=len)
+                if len(temp_lcc) / N > self.in_sample_proportion / 2:
+                    self.in_sample_idx = temp_lcc
+                    less_than_half = False
+                c += 1
+            if c == self.connected_attempts:
+                msg = (
+                    "Induced subgraph is not fully connected. Attempted to find connected"
+                    + " induced subgraph {} times. Results may not be optimal.".format(
+                        self.connected_attempts
+                    )
+                    + " Try increasing proportion of in sample vertices."
+                )
+                warnings.warn(msg, UserWarning)
+
+        return
+
+    def predict(self, X, ids=None):
+        """
+        Embed out of sample vertices.
+
+        Parameters
+        ----------
+        X : array_like, shape (m, n)
+            m stacked similarity vectors, where the jth entry of the ith row corresponds to
+            the similarity of the ith out of sample observation to the jth in sample
+            observation.
+        ids : array_like, length m
```

Comment:
I'd change this to `y` simply for consistency with sklearn and the rest of GraSPy

Reply:
hesitant to do this, `y` typically denotes a list of targets, whereas `ids` here is list node names for the out of sample vertices.

i think `predict` here is a stretch use of the function name, but could not think of something better.